File: android/chat-ui-compose/src/main/kotlin/jp/co/altive/chat/ChatMessageCard.kt
package jp.co.altive.chat

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/** メッセージカードの意味的な外観。 */
@Immutable
enum class ChatMessageCardStyle {
  /** 誕生日や記念日などのお祝いに使う外観。 */
  Celebration,
}

/** アプリが構築した内容を表示する汎用メッセージカード。 */
@Composable
fun ChatMessageCard(
  style: ChatMessageCardStyle,
  isOwnMessage: Boolean,
  accessibilityLabel: String,
  modifier: Modifier = Modifier,
  theme: ChatRoomTheme = ChatRoomTheme.fanely(),
  header: @Composable ColumnScope.() -> Unit,
  content: @Composable ColumnScope.() -> Unit,
  footer: (@Composable ColumnScope.() -> Unit)? = null,
) {
  val shape = RoundedCornerShape(
    topStart = 20.dp,
    topEnd = 20.dp,
    bottomStart = if (isOwnMessage) 20.dp else 8.dp,
    bottomEnd = if (isOwnMessage) 8.dp else 20.dp,
  )
  val background = when (style) {
    ChatMessageCardStyle.Celebration -> Brush.linearGradient(
      listOf(theme.celebrationCardBackgroundStart, theme.celebrationCardBackgroundEnd),
    )
  }

  Box(
    modifier = modifier
      .clip(shape)
      .background(background)
      .border(1.dp, theme.celebrationCardBorder, shape)
      .semantics(mergeDescendants = true) { contentDescription = accessibilityLabel },
  ) {
    CelebrationCardDecoration(
      accent = theme.celebrationCardAccent,
      modifier = Modifier.matchParentSize(),
    )
    CompositionLocalProvider(LocalContentColor provides theme.celebrationCardForeground) {
      Column(Modifier.padding(16.dp)) {
        header()
        Spacer(Modifier.size(8.dp))
        content()
        if (footer != null) {
          Spacer(Modifier.size(12.dp))
          footer()
        }
      }
    }
  }
}

@Composable
private fun CelebrationCardDecoration(
  accent: androidx.compose.ui.graphics.Color,
  modifier: Modifier = Modifier,
) {
  Canvas(modifier) {
    drawCircle(
      color = accent.copy(alpha = .2f),
      radius = 5.dp.toPx(),
      center = Offset(size.width * .12f, size.height * .18f),
    )
    drawCircle(
      color = accent.copy(alpha = .16f),
      radius = 3.5.dp.toPx(),
      center = Offset(size.width * .88f, size.height * .3f),
    )
    rotate(degrees = 38f, pivot = Offset(size.width * .78f, size.height * .82f)) {
      drawRect(
        color = accent.copy(alpha = .18f),
        topLeft = Offset(size.width * .78f - 7.dp.toPx(), size.height * .82f - 2.dp.toPx()),
        size = androidx.compose.ui.geometry.Size(14.dp.toPx(), 4.dp.toPx()),
      )
    }
  }
}

File: lib/src/chat_message_card.dart
import 'package:flutter/material.dart';

import 'models/altive_chat_room_theme.dart';

/// メッセージカードの意味的な外観。
enum ChatMessageCardStyle {
  /// 誕生日や記念日などのお祝いに使う外観。
  celebration,
}

/// アプリが構築した内容を表示する汎用メッセージカード。
class ChatMessageCard extends StatelessWidget {
  /// 汎用メッセージカードを作成する。
  const ChatMessageCard({
    super.key,
    required this.style,
    required this.isOwnMessage,
    required this.accessibilityLabel,
    required this.header,
    required this.content,
    this.footer,
    this.theme = const AltiveChatRoomTheme(),
  });

  /// カードの意味的な外観。
  final ChatMessageCardStyle style;

  /// ログインユーザーが送信したメッセージかどうか。
  final bool isOwnMessage;

  /// カード用途を説明するローカライズ済み読み上げラベル。
  final String accessibilityLabel;

  /// 見出しや宛名を表示するWidget。
  final Widget header;

  /// 本文を表示するWidget。
  final Widget content;

  /// 補足表示を置く任意のWidget。
  final Widget? footer;

  /// カードへ適用するチャットテーマ。
  final AltiveChatRoomTheme theme;

  @override
  Widget build(BuildContext context) {
    final colorScheme = Theme.of(context).colorScheme;
    final borderRadius = BorderRadiusDirectional.only(
      topStart: const Radius.circular(20),
      topEnd: const Radius.circular(20),
      bottomStart: Radius.circular(isOwnMessage ? 20 : 8),
      bottomEnd: Radius.circular(isOwnMessage ? 8 : 20),
    ).resolve(Directionality.of(context));
    final backgroundStart =
        theme.celebrationCardBackgroundStart ??
        colorScheme.tertiaryContainer.withValues(alpha: .62);
    final backgroundEnd =
        theme.celebrationCardBackgroundEnd ??
        colorScheme.secondaryContainer.withValues(alpha: .5);
    final border =
        theme.celebrationCardBorder ??
        colorScheme.tertiary.withValues(alpha: .42);
    final foreground =
        theme.celebrationCardForeground ?? colorScheme.onTertiaryContainer;
    final accent = theme.celebrationCardAccent ?? colorScheme.tertiary;

    return Semantics(
      container: true,
      label: accessibilityLabel,
      child: DecoratedBox(
        decoration: BoxDecoration(
          gradient: switch (style) {
            ChatMessageCardStyle.celebration => LinearGradient(
              begin: AlignmentDirectional.topStart,
              end: AlignmentDirectional.bottomEnd,
              colors: [backgroundStart, backgroundEnd],
            ),
          },
          border: Border.all(color: border),
          borderRadius: borderRadius,
        ),
        child: ClipRRect(
          borderRadius: borderRadius,
          child: Stack(
            children: [
              Positioned.fill(
                child: ExcludeSemantics(
                  child: CustomPaint(
                    painter: _CelebrationCardDecorationPainter(accent),
                  ),
                ),
              ),
              Padding(
                padding: const EdgeInsets.all(16),
                child: IconTheme.merge(
                  data: IconThemeData(color: foreground),
                  child: DefaultTextStyle.merge(
                    style: TextStyle(color: foreground),
                    child: Column(
                      mainAxisSize: MainAxisSize.min,
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        header,
                        const SizedBox(height: 8),
                        content,
                        if (footer != null) ...[
                          const SizedBox(height: 12),
                          footer!,
                        ],
                      ],
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _CelebrationCardDecorationPainter extends CustomPainter {
  const _CelebrationCardDecorationPainter(this.accent);

  final Color accent;

  @override
  void paint(Canvas canvas, Size size) {
    canvas
      ..drawCircle(
        Offset(size.width * .12, size.height * .18),
        5,
        Paint()..color = accent.withValues(alpha: .2),
      )
      ..drawCircle(
        Offset(size.width * .88, size.height * .3),
        3.5,
        Paint()..color = accent.withValues(alpha: .16),
      )
      ..save()
      ..translate(size.width * .78, size.height * .82)
      ..rotate(38 * 3.141592653589793 / 180)
      ..drawRRect(
        RRect.fromRectAndRadius(
          const Rect.fromLTWH(-7, -2, 14, 4),
          const Radius.circular(2),
        ),
        Paint()..color = accent.withValues(alpha: .18),
      )
      ..restore();
  }

  @override
  bool shouldRepaint(_CelebrationCardDecorationPainter oldDelegate) =>
      oldDelegate.accent != accent;
}
